### PR TITLE
feat: use github graphql api for commits

### LINKED ISSUE #1200: Signing commits made by release-plz bot
## Motivations

Hey, I've seen the #956 which added the `Signed-off-by: github-actions[bot] <41898282+github-actions[bot]@users.noreply.github.com>` to the commit message. This however doesn't alone result in `verified` status on github. We have a repository setting which requires all commits to be signed and thus we cannot merge the release PR.

For comparison bot's commit and my commit
image

And the merge button is red with a message:

> Merging is blocked
> The base branch requires all commits to be signed. [Learn more about signing commits.](https://docs.github.com/articles/about-gpg/)

Is there a possibility for getting a verified status for release-plz bot?

- Would you like to implement this feature? [y/n]

## Solution

It looks like the dependabot has a power to get verified status, but I'm not sure if it isn't kinda specially supported by github.

image

### PR DESCRIPTION
**Pull request by zvolin**
closes #1200.

This PR adds a new flag '--api-commit' to `release-pr` command. If present, a release commit will not be created in local copy and pushed. Instead the changes will be encoded with base64 or just marked as deleted and the commit will happen using [github's graphql api](https://docs.github.com/en/graphql), namely `createCommitOnBranch`.

The main benefit is that the commits created using the graphql api will be associated with the token owner. This allows for better flexibility in who is the author, allowing things like getting [Verified](https://docs.github.com/en/authentication/managing-commit-signature-verification/about-commit-signature-verification) status.

This is currently an mvp. Let me know if you like the idea and I can continue working on docs and tests and polish what's needed.

### CODE REVIEW

**Comment by zvolin** [comment]
I've managed to successfully create this PR https://github.com/zvolin/test-release-plz/pull/3

**Comment by MarcoIeni** [comment]
Sorry if I haven't reviewed this yet, I will get to it asap!

**Comment by zvolin** [comment]
Thanks for checking :+1: 

As I mentioned above, I haven't fully committed into the change yet and wanted to first know your thoughts about approach. If you like it.

**Comment by MarcoIeni** [comment]
thanks for digging into this. What's the con of removing this flag and making this the default behavior?

**Comment by zvolin** [comment]
I haven't checked gitea and gitlab api for equivalents but it's likely it's github specific

**Comment by MarcoIeni** [comment]
> I haven't checked gitea and gitlab api for equivalents but it's likely it's github specific

yes, I mean that we could use this approach with the GitHub backend.

> Also I haven't used any graphql Rust crate because they require downloading and hardcoding whole graphql Schema so that all types are generated. I thought for a single api hit, the string manipulation will be simpler

Yes, that's totally fine :)

**Comment by zvolin** [comment]
hmmm if we gonna use it for github by default, I would need to check what happens if someone tries that with the default GITHUB_TOKEN on CI

**Comment by zvolin** [comment]
so should I proceed with the impl?

**Comment by zvolin** [comment]
Linking the previous example commit because I was uncareful and the PR was force-pushed with the regular release-plz one.
The commit created by this change is there: https://github.com/zvolin/test-release-plz/commit/23053bfbd061c8cb1ec501a357e36f105d75e75e

**Comment by zvolin** [comment]
Also worth noting that the verified status doesn't depend on owner's GPG key (how could it) and it is rather verified by github's (some kind of one-time?) signature:
image

**Comment by zvolin** [comment]
> hmmm if we gonna use it for github by default, I would need to check what happens if someone tries that with the default GITHUB_TOKEN on CI

This actually looks good, everything seems to work: https://github.com/zvolin/test-release-plz/pull/4

**Comment by zvolin** [comment]
I think we can make it a default

**Comment by MarcoIeni** [comment]
> so should I proceed with the impl?

I just want to make sure that the graphql api is the simplest way to obtain the "verified" badge. 🤔 
I would like to find out if there's a way of obtaining it by committing from the cli, which is simpler, but you probably researched this already 👍

**Comment by zvolin** [comment]
Hmmm yes I was researching that however I'll appreciate double check. Here is [some write-up (https://github.com/Nautilus-Cyberneering/pygithub/blob/main/docs/how_to_sign_automatic_commits_in_github_actions.md) from 3y ago.
Also https://github.com/google-github-actions/release-please-action/issues/399.

In short, I think the alternative to API is to have a second github account for bot with gpg key set and to also pass the signing key as a secret to the workflow, configure the author and email to that second account and then commit. I haven't researched the github apps tho

**Comment by MarcoIeni** [comment]
I did some research and I think your approach is the best one because it avoids storing GPG secrets in the repository.
If it works well, I would like to make it the default for the github backend.

**Comment by MarcoIeni** [comment]
I don't think there should be support for the old behavior unless people complain and they provide a reason.

We could update the release-pr command docs, yes.

For tests it's difficult, because we don't test github api.

**Comment by zvolin** [comment]
> There's also a small merge conflict right now. Can you resolve it?

Sure.

> For tests it's difficult, because we don't test github api.

I'll add some unit tests then and call it a day.

I'm not sure I can get this ready over the weekend as I'll be in travel, we'll see.

**Comment by zvolin** [comment]
I removed the API_ENDPOINT constant in favor of taking base_url of the GitClient. I thought it might be useful at some point due to the remote mock in existing tests. Maybe not even useful but if someone in future tries to write some tests will not be surprised that he hits github api

**Comment by zvolin** [comment]
Sooo....
I wanted to write some tests and I gave up. I had a test for this newly added `get_graphql_endpoint` that checks if it fails for other git upstreams and I wanted to write one which checks for correct inclusion of unchanged, changed, type_changed, added and deleted files in repo. However, even the output of stuff like `format_additions` cannot be parsed as a json (graphql doesn't quote key names, like `path: "foo"`) and the test ended up being really cumbersome to the point I think it's not worth it

**Comment by zvolin** [comment]
actually after writing that I noticed that the domain check in get_graphql_endpoint defeats it's purpose so removed also that

**Comment by zvolin** [comment]
Updated the docs but please check those carefully, English is not one of my advantages :upside_down_face: This is ready by me

**Comment by zvolin** [comment]
> Do you expect someone to miss the old behavior for any reason?

just noticed I haven't answered that. So I think I've seen some complaints that the changes were so big that encoding them as base64 and pushing was overflowing some (i think 10mb but can't be sure as don't remeber source) constrain on graphql api. However, I don't expect that to happen with `release-plz release-pr`.

File: crates/release_plz_core/src/git/github_graphql.rs [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+              deletions: {deletions},
+              additions: {additions}
+            }}
+          }}) {{
+            commit {{
+              author {{
+                name,
+                email
+              }}
+            }}
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/git/github_graphql.rs:112]: Is this the result returned by the mutation? If yes, do we need it?
I don't think we are reading it.

File: crates/release_plz_core/src/git/github_graphql.rs [hunk 2]
```diff
@@ -0,0 +1,168 @@
...
+
+// format a graphql query to create commit on branch
+async fn format_commit_query(
+    owner_and_repo: &str,
+    branch: &str,
+    message: &str,
+    current_head: &str,
+    deletions: &[impl AsRef<Path>],
+    additions: &[impl AsRef<Path>],
+    repo_dir: impl AsRef<Path>,
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/git/github_graphql.rs:88]: can you collect all these parameters into a struct? E.g. `GitHubCommit`

File: crates/release_plz_core/src/git/github_graphql.rs [hunk 3]
```diff
@@ -0,0 +1,168 @@
...
+    repo.changes(|line| !line.starts_with("T ") && !line.starts_with("D "))
+}
+
+// get the list of removed files in repository
+fn removed_files(repo: &Repo) -> Result<Vec<String>> {
+    repo.changes(|line| line.starts_with("D "))
+}
+
+// format a graphql query to create commit on branch
+async fn format_commit_query(
```
Comment (MarcoIeni) [inline crates/release_plz_core/src/git/github_graphql.rs:81]: Can you add a unit test for this where you test that the correct query is returned?
No need to send the query over the network, let's just test that the query is constructed correctly.

For the directories deletions and additions, you could create some files in a temporary directory like we do [here](https://github.com/zvolin/release-plz/blob/32bd7aa6e99475d14a29fb7766076a817b5a331d/crates/release_plz_core/src/copy_dir.rs#L108)

**Review by MarcoIeni — CHANGES_REQUESTED** [review]
I did some small refinements, but overall looks great! 🚀
I just requested a few minor changes. 🙏

File: crates/release_plz_core/src/git/github_graphql.rs (hunk 1 above)
Comment (zvolin) [reply crates/release_plz_core/src/git/github_graphql.rs:111]: it is a selection of some data from what the mutation returns. In graphql you must have some selection of fields up to some primitives. Eg.
```
commit {{ }}
```
would complain that you didn't select any subfields of commit.

It is not read anywhere but there must be some selection.

File: crates/release_plz_core/src/git/github_graphql.rs (hunk 3 above)
Comment (zvolin) [reply crates/release_plz_core/src/git/github_graphql.rs]: done

File: crates/release_plz_core/src/git/github_graphql.rs (hunk 2 above)
Comment (zvolin) [reply crates/release_plz_core/src/git/github_graphql.rs]: done

File: crates/release_plz_core/src/git/github_graphql.rs (hunk 1 above)
Comment (MarcoIeni) [reply crates/release_plz_core/src/git/github_graphql.rs]: oh, right! Sounds good then :)

**Comment by MarcoIeni** [comment]
image
Just tested it and got the juicy "verified" badge. Thanks for reporting the issue, researching how to solve it, submitting a POC, and refining the PR after seeking feedback. Awesome job! 🙌

**Review by MarcoIeni — APPROVED** [review]
Thanks ❤️

**Comment by zvolin** [comment]
Thanks too :)

**Pull request merged**